python/pyspark/sql/types.py

```diff
@@ -1619,11 +1619,39 @@ def to_arrow_type(dt):
         arrow_type = pa.decimal(dt.precision, dt.scale)
     elif type(dt) == StringType:
         arrow_type = pa.string()
+    elif type(dt) == DateType:
+        arrow_type = pa.date32()
+    elif type(dt) == TimestampType:
+        # Timestamps should be in UTC, JVM Arrow timestamps require a timezone to be read
+        arrow_type = pa.timestamp('us', tz='UTC')
     else:
         raise TypeError("Unsupported type in conversion to Arrow: " + str(dt))
     return arrow_type
 
 
+def _check_dataframe_localize_timestamps(df):
+    """ Convert timezone aware timestamps to timezone-naive in local time
+    """
+    from pandas.types.common import is_datetime64tz_dtype
+    for column, series in df.iteritems():
+        # TODO: handle nested timestamps?
+        if is_datetime64tz_dtype(series.dtype):
+            df[column] = series.dt.tz_convert('tzlocal()').dt.tz_localize(None)
+    return df
+
+
+def _check_series_convert_timestamps_internal(s):
+    """ Convert a tz-naive timestamp in local tz to UTC normalized for Spark internal storage
+    """
+    from pandas.types.common import is_datetime64_dtype
+    # TODO: handle nested timestamps?
+    if is_datetime64_dtype(s.dtype):
+        # NOTE: convert to 'us' with astype here, unit is ignored in `from_pandas` see ARROW-1680
+        return s.dt.tz_localize('tzlocal()').dt.tz_convert('UTC').values.astype('datetime64[us]')
```

Comment:
Can you create a follow up JIRA to fix this as soon as Arrow 0.8.0 lands? I will make sure that this gets fixed

Reply:
Sure, there are a number of thinks marked to be fixed when an upgrade happens.  I'm thinking of first collecting them all in a JIRA first to help spur discussion for upgrading Arrow.